skopt/acquisition.py

```diff
@@ -264,3 +267,119 @@ def gaussian_ei(X, model, y_opt=0.0, xi=0.01, return_grad=False):
+        It should have a ``return_std`` parameter that returns the standard
+        deviation.
+
+    * `space` [Space instance]:
+        This is required to sample representer points.
+
+    * `n_rep_points` [int, default 50]:
+        Number of represter points.
+
+    * `n_trial_points` [int, default 200]:
+        Number of trial points sampled uniformly for each representer point
+        candidate.
+
+    * `n_samples` [int, default 500]:
+        Number of samples to sample from the GP posterior to approximate p_min
+
+    * `random_state` [int, RandomState instance, or None (default)]:
+        For reproducible results.
+    """
+    rng = np.random.RandomState(random_state)
+    n_features = X.shape[1]
+
+    # Step 1: Sample `n_representer_points` from the given nonmeasure.
+    # XXX: Use MCMC sampling to sample from the quadrature measure.
+    n_total = n_rep_points*n_trial_points
+    rep_cands = space.transform(space.rvs(n_samples=n_total, random_state=rng))
+    rep_points = np.zeros((n_rep_points, n_features))
+
+    if measure == "LCB":
+        quad_func = gaussian_lcb
+
+    elif measure in ["EI", "PI"]:
+        if measure == "EI":
+            quad_func = lambda X, model: -gaussian_ei(X, model)
+        else:
+            quad_func = lambda X, model: -gaussian_pi(X, model)
+
+    for i, startptr in enumerate(range(0, X.shape[0], n_total)):
+        rep_curr = rep_cands[startptr: startptr+n_trial_points]
+        argmin = np.argmin(quad_func(rep_curr, model))
+        rep_points[i] = rep_curr[argmin]
+
+    # # Step 2: Calculate p_min for each of these represnter points using
+    # # Monte-Carlo approximation.
+    # sampled_f = model.sample_y(
+    #     X_candidates, n_samples=n_samples, random_state=rng)
+    # p_min = (np.bincount(np.argmax(sampled_f, axis=0)) + 1) / (n_candidates + float(n_samples))
+    # base_entropy = entropy(p_min)
+
+    X_all = np.vstack((rep_points, X))
+    X_all_mean, X_all_cov = model.predict(X_all, return_cov=True)
+
+    # Store mu(X_i) and Sigma(X_i, X_j) of all the representer points based on
+    # previous evaluations.
+    init_cov = X_all_cov[:n_rep_points, :n_rep_points]
+    init_mean = X_all_mean[:n_rep_points]
+
+    # For each candidate point calculate change in p_min of all the represnter
+    # points and hence change in entropy (information gain)
+    inf_gain = np.zeros(X.shape[0])
+    for cand_ind in range(n_rep_points, n_rep_points + X.shape[0]):
+
+        # delta(cov(X_rep)) = \Sigma(X_rep, X_cand) \Sigma^-1(X_cand, X_cand) \Sigma(X_cand, X_rep)
+        # Note that \Sigma(X_cand, X_cand) is a scalar since we only require the next
+        # best evaluation.
+        cov_rep_X = X_all_cov[cand_ind][: n_rep_points]
+        cov_rep_X_row = np.reshape(cov_rep_X, (-1, 1))
+        cov_rep_X_col = np.reshape(cov_rep_X, (1, -1))
+        cov_cand_inv = 1.0 / X_all_cov[cand_ind, cand_ind]
+        cov_delta = cov_cand_inv * np.dot(cov_rep_X_row, cov_rep_X_col)
```

Comment:
Yes, you are right. There should be a negative sign. Do you know the approach to derive it analytically? Here is a numeric approach.

```python
gpr = GaussianProcessRegressor(n_restarts_optimizer=5)
X = [[1.0], [2.0], [3.0], [4.0], [5.0]]
y = [1.0, 4.0, 9.0, 16.0, 25.0]
gpr.fit(X, y)

eval_X = np.linspace(1.0, 5.0, 30.0).reshape(-1, 1)
_, init_cov = gpr.predict(eval_X, return_cov=True)

new_X = [[1.5], [2.5], [3.5]]
new_y = y + [t**2 for x in new_X for t in x]

all_X = np.vstack((eval_X, new_X))
_, all_cov = gpr.predict(all_X, return_cov=True)
cross_cov = all_cov[-3:, :-3]
delta_cov = np.dot(cross_cov.T, np.dot(np.linalg.inv(all_cov[-3:, -3:]), cross_cov))

X_after = np.vstack((X, new_X))
gpr.fit(X_after, new_y)
_, new_cov = gpr.predict(eval_X, return_cov=True)
act_delta = new_cov - init_cov

act_delta[-2]
array([ 0.        ,  0.00118236,  0.00206068,  0.00254834,  0.00259283,
        0.00218859,  0.00138669,  0.00029841, -0.00091064, -0.00203781,
       -0.00287196, -0.00323024, -0.00299673, -0.00215471, -0.00080388,
        0.00084434,  0.00249582,  0.00382406,  0.00453331,  0.00442178,
        0.00343106,  0.00167007, -0.00059326, -0.00297201, -0.00502493,
       -0.00633262, -0.00657042, -0.0055635 , -0.00331457,  0.        ])
delta_cov[-2]
array([  0.00000000e+00,  -1.18235670e-03,  -2.06067792e-03,
        -2.54833878e-03,  -2.59283482e-03,  -2.18858735e-03,
        -1.38668602e-03,  -2.98406054e-04,   9.10643899e-04,
         2.03780813e-03,   2.87196229e-03,   3.23024283e-03,
         2.99672751e-03,   2.15471226e-03,   8.03883150e-04,
        -8.44338624e-04,  -2.49581620e-03,  -3.82405586e-03,
        -4.53330821e-03,  -4.42177769e-03,  -3.43106330e-03,
        -1.67006787e-03,   5.93257256e-04,   2.97201330e-03,
         5.02492645e-03,   6.33262299e-03,   6.57041884e-03,
         5.56350449e-03,   3.31456744e-03,   6.71512219e-17])

```